icu4c/source/i18n/measfmt.cpp

```diff
@@ -770,71 +715,96 @@ static void appendRange(
-    // is 9.35s and the formatted time is 0:00:09 then 9.35 becomes 09.35
-    // and replacing yields 0:00:09.35
-    if (smallestFieldPosition.getBeginIndex() != 0 ||
-            smallestFieldPosition.getEndIndex() != 0) {
-        appendRange(draft, 0, smallestFieldPosition.getBeginIndex(), appendTo);
-        appendRange(
-                smallestAmountFormatted,
-                0,
-                intFieldPosition.getBeginIndex(),
-                appendTo);
-        appendRange(
-                draft,
-                smallestFieldPosition.getBeginIndex(),
-                smallestFieldPosition.getEndIndex(),
-                appendTo);
-        appendRange(
-                smallestAmountFormatted,
-                intFieldPosition.getEndIndex(),
-                appendTo);
-        appendRange(
-                draft,
-                smallestFieldPosition.getEndIndex(),
-                appendTo);
-    } else {
-        appendTo.append(draft);
+    const NumberFormat *numberFormatter = numberFormat->get();
+    auto* numberFormatter2 = (NumberFormat*) numberFormatter->clone();
+    numberFormatter2->setMinimumIntegerDigits(2);
+
+    FormattedStringBuilder fsb;
+
+    UBool protect = FALSE;
+    const int32_t patternLength = pattern.length();
+    for (int32_t i = 0; i < patternLength; i++) {
+        char c = pattern[i];
+
+        // Also set the proper field in this switch
+        // We don't use DateFormat.Field because this is not a date / time, is a duration.
+        double value = 0;
+        switch (c) {
+            case 'H': value = hours; break;
+            case 'm': value = minutes; break;
+            case 's': value = seconds; break;
+        }
+
+        // There is not enough info to add Field(s) for the unit because all we have are plain
+        // text patterns. For example in "21:51" there is no text for something like "hour",
+        // while in something like "21h51" there is ("h"). But we can't really tell...
+        switch (c) {
+            case 'H':
+            case 'm':
+            case 's':
+                if (protect) {
+                    fsb.append(UnicodeString(c), (FormattedStringBuilder::Field) NULL, status);
+                } else {
+                    UnicodeString tmp;
+                    if ((i + 1 < patternLength) && pattern[i + 1] == c) { // doubled
+                        numberFormatter2->format(value, tmp, status);
+                        i++;
+                    } else {
+                        numberFormatter->format(value, tmp, status);
+                    }
+                    // TODO: Use proper Field
+                    fsb.append(tmp, (FormattedStringBuilder::Field) NULL, status);
+                }
+                break;
+            case '\'':
+                // '' is escaped apostrophe
+                if ((i + 1 < patternLength) && pattern[i + 1] == c) {
+                    fsb.append(UnicodeString(c), (FormattedStringBuilder::Field) NULL, status);
```

Comment:
ditto

Reply:
Can't use appendCodePoint because I iterate over code units (the `[]` operator returns `char16_t`)

Options:
```
int patternLen = pattern.countChar32();
for (int i = 0; i < patternLen; i++) {
    UChar32 c = pattern.char32At(i);
    ...
    fsb.appendCodePoint(c, (FormattedStringBuilder::Field) NULL, status);
}
```
or
```
int patternLen = pattern.length();
for (int i = 0; i < patternLen; i++) {
    char16_t c = pattern[i];
    fsb.append(UnicodeString(c), (FormattedStringBuilder::Field) NULL, status);
}
```

`UnicodeString(c)` might be expensive.
But `char32At` is also expensive, because it has to iterate the whole string, for each get.
Maybe convert the whole string to a `UChar32` array.

For the first option (`char32At`) the documentation says _"Return the code point that contains the code unit at offset `<tt>`offset`</tt>`."_
I find that a bit confusing. Is that the offset in code points, or code units?
Trying it seems to indicate that it is code point offset (as the name suggests). The doc makes things more confusing instead of more clear...

Reply:
> Is that the offset in code points, or code units?

The offset is always in code units.

Reply:
Done.